
# Phologenomics

## Introduction



All nine bash scripts are tested in Linux systems. Their functions are described below.





## Requirements

Some bioinformatic tools are neccessary for above scripts. Most of them are recommended to be added into the environmental paths. Softwares, versions and source are listed below.

BBTools v38.32 (https://sourceforge.net/projects/bbmap/) 

Minia v3.2.4 (https://github.com/GATB/minia) 

Redundans v0.14c (https://github.com/lpryszcz/redundans) 

Minimap2 v2.12 (https://github.com/lh3/minimap2) 

Samtools v1.9 (http://www.htslib.org/) 

BESST v2.2.8 (https://github.com/ksahlin/BESST) 

GapCloser v1.12 (http://soap.genomics.org.cn/) 

BUSCO v3.0.2 (http://busco.ezlab.org/) 

FASconCAT-G v1.04 (https://github.com/PatrickKueck/FASconCAT-G) 

PHYLUCE v1.6.6 (http://phyluce.readthedocs.io/en/latest/index.html) 

faToTwoBit (http://hgdownload.soe.ucsc.edu/admin/exe/) 

ART-20160605 (https://www.niehs.nih.gov/research/resources/software/biostatistics/art/index.cfm) 

Stampy v1.0.32 (http://www.well.ox.ac.uk/project-stampy) 

PhyKIT (https://github.com/JLSteenwyk/PhyKIT)

R v4.1.2 (https://www.r-project.org/) 

TransDecoder v5.5.0 (https://github.com/TransDecoder/TransDecoder)

TreeShrink v1.3.8 (https://github.com/uym2/TreeShrink/tree/v1.3.8)

MAGUS (https://github.com/vlasmirnov/MAGUS) 

MAFFT v7.407 (https://mafft.cbrc.jp/alignment/software/) 

ClipKIT (https://github.com/JLSteenwyk/ClipKIT) 

trimAl v1.4.1 (http://trimal.cgenomics.org/) 

IQ-TREE v2.1.3 (https://github.com/iqtree/iqtree2) 

ASTRAL-III (https://github.com/smirarab/ASTRAL) 

pbmpi-1.8c (https://github.com/bayesiancook/pbmpi)

## User manual


